core/unix/os.c

```diff
@@ -3389,20 +3391,26 @@ emulate_app_brk(dcontext_t *dcontext, byte *new_val)
     ASSERT(DYNAMO_OPTION(emulate_brk));
     LOG(THREAD, LOG_HEAP, 2, "%s: cur=" PFX ", requested=" PFX "\n", __FUNCTION__,
         app_brk_cur, new_val);
+    byte *request_val = new_val;
     new_val = (byte *)ALIGN_FORWARD(new_val, PAGE_SIZE);
     if (new_val == NULL || new_val == app_brk_cur ||
         /* Not allowed to shrink below original base */
         new_val < app_brk_map) {
         /* Just return cur val */
+        if (new_val == app_brk_cur) {
+            app_brk_request = request_val;
```

Comment:
Please add comments. The logic is hard to follow. This is not "Just return cur_val" as the comment says: this could return a different value if the app requests an increment still within the same page, right? This if() does not match that comment at all.